### PR TITLE
Add {create|update}InterfaceDeclaration methods

### LINKED ISSUE #15497: Not possible to create an Interface definition in synthetic AST
**TypeScript Version:**  2.3.2

**Code**

```ts
// A *self-contained* demonstration of the problem follows...

// There doesn't seem to be an API to create an Interface in the AST programmatically
const _implementsClause = ts.createHeritageClause(ts.SyntaxKind.ImplementsKeyword, [
    ts.createExpressionWithTypeArguments([], ts.createIdentifier('MyInterface'))
]);
const _exportModifier = ts.createToken(ts.SyntaxKind.ExportKeyword);
const _classExpression = ts.createClassExpression([_exportModifier], 'MyClass', [], [_implementsClause], []);
const _classStatement = ts.createStatement(_classExpression);

let src = ts.createSourceFile('output.ts', '', ts.ScriptTarget.ES5, false, ts.ScriptKind.TS);
src = ts.updateSourceFileNode(src, [_classStatement]);
```

**Expected behavior:**
There should be an API for creating an Interface definition in a synthetic AST.

**Actual behavior:**
Doesn't seem like the API exists.

### PR DESCRIPTION
**Pull request by phated**
Fixes #15497

### CODE REVIEW

File: src/compiler/factory.ts [hunk 1]
```diff
@@ -1440,6 +1440,28 @@ namespace ts {
             : node;
     }
 
+    export function createInterfaceDeclaration(decorators: Decorator[] | undefined, modifiers: Modifier[] | undefined, name: string | Identifier | undefined, typeParameters: TypeParameterDeclaration[] | undefined, heritageClauses: HeritageClause[], members: TypeElement[]) {
+        const node = <InterfaceDeclaration>createSynthesizedNode(SyntaxKind.InterfaceDeclaration);
+        node.decorators = asNodeArray(decorators);
```
Comment (mohsen1) [inline src/compiler/factory.ts]: Interfaces don't have decorators AFAIK.

Comment (phated) [inline src/compiler/factory.ts]: Interesting that it didn't give me a type like some of the other members.

Comment (rbuckton) [inline src/compiler/factory.ts]: We parse decorators for interfaces but only to report errors currently. We do plan to support "ambient" decorators (compiler only) at some point in the future, so it is better to have them included in the signature.

File: src/compiler/factory.ts [hunk 2]
```diff
@@ -1440,6 +1440,28 @@ namespace ts {
             : node;
     }
 
+    export function createInterfaceDeclaration(decorators: Decorator[] | undefined, modifiers: Modifier[] | undefined, name: string | Identifier | undefined, typeParameters: TypeParameterDeclaration[] | undefined, heritageClauses: HeritageClause[], members: TypeElement[]) {
```
Comment (rbuckton) [inline src/compiler/factory.ts]: An interface must always have a name, so `name` should be `string | Identifier`. `undefined` is not permitted.

File: src/compiler/factory.ts [hunk 3]
```diff
@@ -1440,6 +1440,28 @@ namespace ts {
...
+        node.decorators = asNodeArray(decorators);
+        node.modifiers = asNodeArray(modifiers);
+        node.name = asName(name);
+        node.typeParameters = asNodeArray(typeParameters);
+        node.heritageClauses = asNodeArray(heritageClauses);
+        node.members = createNodeArray(members);
+        return node;
+    }
+
+    export function updateInterfaceDeclaration(node: InterfaceDeclaration, decorators: Decorator[] | undefined, modifiers: Modifier[] | undefined, name: Identifier | undefined, typeParameters: TypeParameterDeclaration[] | undefined, heritageClauses: HeritageClause[], members: TypeElement[]) {
```
Comment (rbuckton) [inline src/compiler/factory.ts]: An interface must always have a name, so `name` should be `Identifier` here. `undefined` is not permitted.

Comment (rbuckton) [inline src/compiler/factory.ts]: `heritageClauses` can be `undefined`, so the type should be `HeritageClause[] | undefined`.

File: src/compiler/factory.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/factory.ts]: `heritageClauses` can be `undefined`, so the type should be `HeritageClause[] | undefined`.

**Comment by phated** [comment]
Updated

**Pull request merged**